src/gluonts/dataset/loader.py

```diff
@@ -11,186 +11,270 @@
-        acquiring data from iterators.
-        Note that using large prefetching batch will provide smoother bootstrapping performance,
-        but will consume more shared_memory. Using smaller number may forfeit the purpose of using
-        multiple worker processes, try reduce `num_workers` in this case.
-        By default it defaults to `num_workers * 2`.
-    cyclic
-        Indicates whether the dataset is traversed potentially multiple times.
-    shuffle_buffer_length
-        The length of the buffer used to do pseudo shuffle.
-        If not None, the loader will perform pseudo shuffle when generating batches.
-        Note that using a larger buffer will provide more randomized batches, but will make the job require a bit
-        more time to be done.
+    A wrapper class which takes a serialized iterator as an input and generates a
+    pseudo randomized iterator using the same elements from the input iterator.
     """
 
+    def __init__(self, base_iterator: Iterator, shuffle_buffer_length: int):
+        self.shuffle_buffer: list = []
+        self.shuffle_buffer_length = shuffle_buffer_length
+        self.base_iterator = base_iterator
+        self.base_iter_finished = False
+
+    def __next__(self):
+        # if the buffer is empty, fill the buffer first.
+        # (Should only executed in the first round)
+        if not self.shuffle_buffer:
+            self.shuffle_buffer = list(
+                itertools.islice(
+                    self.base_iterator, self.shuffle_buffer_length
+                )
+            )
+
+        # If buffer still empty, means all elements used, return a signal of
+        # end of iterator
+        if not self.shuffle_buffer:
+            raise StopIteration
+
+        # Choose an element at a random index and yield it and fill it with
+        # the next element in the sequential generator
+        idx = random.randint(0, len(self.shuffle_buffer) - 1)
+        next_sample = self.shuffle_buffer[idx]
+
+        # Replace the index with the next element in the iterator if the
+        # iterator has not finished. Delete the index otherwise.
+        try:
+            self.shuffle_buffer[idx] = next(self.base_iterator)
+        except StopIteration:
+            del self.shuffle_buffer[idx]
+
+        return next_sample
+
+
+class MultiProcessBatcher(Iterator):
     def __init__(
         self,
-        dataset: Dataset,
-        *,
-        transform: Transformation,
-        cyclic: bool,
-        is_train: bool,
+        base_iterable: Iterable,
         batch_size: int,
-        ctx: mx.Context,
-        dtype: DType = np.float32,
-        num_workers: Optional[int] = None,
-        num_prefetch: Optional[int] = None,
-        shuffle_buffer_length: Optional[int] = None,
-        **kwargs,
-    ) -> None:
+        batchify_fn: Callable,
+        num_workers: int,
+        max_queue_size: Optional[int] = None,
+        todevice_fn: Callable = lambda x: x,
```

Comment:
I think `todevice_fn` is not a good name.

In my experiments I had a pair of `encode() / decode()` functions, which were called just before putting things into the queue, and just after when popping a value from it.